### ISSUE TITLE
[Feature request]: Determine policy which caused rejection

### ISSUE THREAD

**Issue by lonix1** [App-vNext/Polly#2345]
### Is your feature request related to a specific problem? Or an existing feature?

I've spent some time learning how Polly works, and specifically how to combine policies (which is probably a common need in a non-trivial production system).

I've found a pain point in how Polly handles rejections (excuse the pun).

Suppose one combines rate limiting and concurrency control, without queuing (or after reaching the queue limit), and there is a rejection:
- if rate limited: read delay from `RateLimiterRejectedException.RetryAfter` then retry
- if concurrency limited: return to caller with error

To do that one needs to know the source of the rejection.

### Describe the solution you'd like

There are two workarounds, both bad:
- If the `RateLimiterRejectedException.RetryAfter` is null, one can infer that it was concurrency limited (rather than fixed window, sliding window or token bucket). But this is a dangerous assumption as the library could add more policies in the future, or change it's internals.
- One could use a variation of the roundabout approach [shown here](https://github.com/App-vNext/Polly/discussions/1777#discussioncomment-7506465) (though I'm not even certain it works... I couldn't get it to work). Assuming it works, it's very messy and overly complicated. I wouldn't want to maintain such code if I only ever touch Polly code once or twice a year.

The ideal solution is mentioned in that thread, which is to add a new property to the exception, which identifies the source of the rejection.

Perhaps `RateLimiterRejectedException.Source` could be a string, equal to a configurable `Name` property for the policy, or if unset then equal to `policy.GetType().Name`.

### Additional context

Thank you for considering it!

**Comment by peter-csala** [comment]
Just to clarify: which Polly version are we talking about? The `RateLimiterRejectedException` was introduced in the V8 API. V7 API uses _policies_ whereas the V8 uses _strategies_. I educated guess is that you are using the V8 API, can you please confirm it?

**Comment by peter-csala** [comment]
Could you please describe the desired behavior in a bit more detail (how many strategies are chain, in what order, how are they configured, etc.)? Or you could also share an ideal pipeline setup.

**Comment by peter-csala** [comment]
Well, as far as I know there is only a single place where Polly throws `RateLimiterRejectedException`:

https://github.com/App-vNext/Polly/blob/d36058d0ac55ac221c1f0d9405aff932998c045a/src/Polly.RateLimiting/RateLimiterResilienceStrategy.cs#L68

I think we could extend the RLRE with a `Source` property and populate it similarly like the Telemetry event's `Source`:
```none
Resilience event occurred. EventName: 'OnRateLimiterRejected', Source: '(null)/(null)/RateLimiter', Operation Key: '', Result: ''
Resilience event occurred. EventName: 'OnRateLimiterRejected', Source: 'MyPipeline/MyPipelineInstance/MyRateLimiterStrategy', Operation Key: 'MyRateLimitedOperation', Result: ''
```

https://github.com/App-vNext/Polly/blob/d36058d0ac55ac221c1f0d9405aff932998c045a/src/Polly.Extensions/Telemetry/Log.cs#L26-L35

That would require minimal code code on your pipeline setup
```cs
_resiliencePipeline = new ResiliencePipelineBuilder()
    // outer strategy: limit threads
    .AddRateLimiter(new RateLimiterStrategyOptions {
      Name = "ThreadLimiter"
      RateLimiter = args => _rateLimiterConcurrent.AcquireAsync(args.Context, permitCount:1, args.Context.CancellationToken),
    })
    // inner strategy: limit requests (per second and per day)
    .AddRateLimiter(new RateLimiterStrategyOptions {
      Name = "RequestLimiter"
      RateLimiter = args => _rateLimiterChained.AcquireAsync(args.Context, permitCount:1, args.Context.CancellationToken),
    })
    .Build();
```
With this in your hand, the `RateLimiterRejectedException` catch block could branch based on the `Source` with `Contains`/`EndsWith`.

Does it sound good for you?

**Comment by peter-csala** [comment]
@lonix1 Do you want to give it a try and file a PR? 

> Unrelated side issue: is it correct to have the concurrency or rate limiter first?

The ordering should not matter:
- rate limiter controls inbound load 
- concurrency limiter controls outbound load

They are both proactive strategies. If they were reactive then the outer's `ShouldHandle` could be adjusted to handle the inner's thrown exception as well. But that's not the case here.

**Comment by martintmk** [comment]
The proposal looks good to be. The only thing I suggest is too actually use the following class as a source property:

https://github.com/App-vNext/Polly/blob/main/src/Polly.Core/Telemetry/ResilienceTelemetrySource.cs

Basically, to uniquely identify strategy you also need pipeline name and pipeline instance name.

**Comment by peter-csala** [comment]
> PR: I don't know... I'm still new to Polly and there's many things I don't yet grok.

Sure, no problem. @martincostello could you please assign to me this issue?

**Comment by peter-csala** [comment]
Hi @lonix1 we had made the required changes and it will be available as part of the 8.5.0 release. The new property name will be `TelemetrySource` (the `Source` was already defined by the `Exception` type...). It is not a string rather a [`ResilienceTelemetrySource`](https://www.pollydocs.org/api/Polly.Telemetry.ResilienceTelemetrySource.html). This class includes not just the strategy name but information about the pipeline as well.

So, the usage will look like this:

```cs
try 
{
    await pipeline.ExecuteAsync(...);
}
catch (RateLimiterRejectedException ex) when (ex.TelemetrySource.StrategyName?.Contains("ThreadLimiter"))
{
  // Your concurrency limiter related logic
}
```

**Comment by peter-csala** [comment]
Hey @martincostello can we close this issue?
